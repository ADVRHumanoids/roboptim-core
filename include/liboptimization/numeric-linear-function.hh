// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the NumericLinearFunction class.
 */

#ifndef OPTIMIZATION_NUMERIC_LINEAR_FUNCTION_HH
# define OPTIMIZATION_NUMERIC_LINEAR_FUNCTION_HH

# include <liboptimization/linear-function.hh>

namespace optimization
{
  /// Define a quadratic function.
  class NumericLinearFunction : public LinearFunction
  {
  public:
    NumericLinearFunction (const matrix_t&, const vector_t&) throw ();

    virtual vector_t operator () (const vector_t&) const throw ();

    virtual gradient_t gradient (const vector_t&, int) const throw ();
    virtual jacobian_t jacobian (const vector_t&) const throw ();

    virtual std::ostream& print (std::ostream&) const throw ();

  private:
    matrix_t a_;
    vector_t b_;
  };

} // end of namespace optimization

#endif //! OPTIMIZATION_QUADRATIC_FUNCTION_HH
