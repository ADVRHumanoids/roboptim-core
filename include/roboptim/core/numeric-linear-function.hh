// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_NUMERIC_LINEAR_FUNCTION_HH
# define ROBOPTIM_CORE_NUMERIC_LINEAR_FUNCTION_HH

# include <roboptim/core/linear-function.hh>

namespace roboptim
{
  /// \addtogroup roboptim_function
  /// @{

  /// \brief Build a linear function from a vector and a matrix.
  ///
  /// Implement a linear function using the general formula:
  /// \f[f(x) = A x + b\f]
  /// where \f$A\f$ and \f$b\f$ are set when the class is instantiated.
  class NumericLinearFunction : public LinearFunction
  {
  public:
    /// \brief Build a linear function from a matrix and a vector.
    ///
    /// See class documentation for A and b definition.
    /// \param A A matrix
    /// \param b b vector
    NumericLinearFunction (const matrix_t& A, const vector_t& b) throw ();
    ~NumericLinearFunction () throw ();

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream&) const throw ();

  protected:
    void impl_compute (result_t& , const vector_t&) const throw ();
    void impl_gradient (gradient_t&, const vector_t&, int = 0) const throw ();
    void impl_jacobian (jacobian_t&, const vector_t&) const throw ();

  private:
    /// \brief A matrix.
    matrix_t a_;
    /// \brief B vector.
    vector_t b_;
  };

  /// Example shows numeric linear function use.
  /// \example numeric-linear-function.cc

  /// @}

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_QUADRATIC_FUNCTION_HH
