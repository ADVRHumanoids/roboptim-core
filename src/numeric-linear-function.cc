// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "debug.hh"

#include <boost/numeric/ublas/io.hpp>
#include <boost/numeric/ublas/matrix.hpp>
#include <boost/numeric/ublas/vector.hpp>

#include <roboptim/core/indent.hh>
#include <roboptim/core/numeric-linear-function.hh>

namespace roboptim
{
  NumericLinearFunction::NumericLinearFunction (const matrix_t& a,
                                                const vector_t& b)
    throw ()
    : LinearFunction (a.size2 (), a.size1 (), "numeric linear function"),
      a_ (a),
      b_ (b)
  {
    assert (b.size () == outputSize ());
  }

  NumericLinearFunction::~NumericLinearFunction () throw ()
  {
  }


  // A * x + b
  void
  NumericLinearFunction::impl_compute (result_t& result,
				       const argument_t& argument)
    const throw ()
  {
    using namespace boost::numeric::ublas;
    noalias (result) = prec_prod (a_, argument) + b_;
  }

  // A
  void
  NumericLinearFunction::impl_jacobian (jacobian_t& jacobian,
					const argument_t&) const throw ()
  {
    jacobian = this->a_;
  }

  // A(i)
  void
  NumericLinearFunction::impl_gradient (gradient_t& gradient,
					const argument_t&,
					size_type idFunction) const throw ()
  {
    for (size_type j = 0; j < inputSize (); ++j)
      gradient[j] = a_ (idFunction, j);
  }

  std::ostream&
  NumericLinearFunction::print (std::ostream& o) const throw ()
  {
    return o << "Numeric linear function" << incindent << iendl
             << "A = " << this->a_ << iendl
             << "B = " << this->b_
             << decindent;
  }

} // end of namespace roboptim
