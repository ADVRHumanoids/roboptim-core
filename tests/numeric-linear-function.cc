// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>
#include <boost/numeric/ublas/io.hpp>

#include <liboptimization/numeric-linear-function.hh>
#include "common.hh"

using namespace optimization;

typedef DummySolver solver_t;

int run_test ()
{
  NumericLinearFunction::vector_t a (5);
  NumericLinearFunction::value_type b = 1.;
  NumericLinearFunction::vector_t x (5);

  a[0] = 1.2;
  a[1] = 3.4;
  a[2] = 5.6;
  a[3] = 7.8;

  NumericLinearFunction f (a, b);

  std::cout << f << std::endl;

  x[0] = 0.1;
  x[1] = 1.2;
  x[2] = 2.3;
  x[3] = 3.4;
  x[4] = 4.5;

  std::cout << "f(x) = " << f (x) << std::endl;
  std::cout << "G(x) = " << f.gradient (x) << std::endl;
  std::cout << "H(x) = " << f.hessian (x) << std::endl;

  return 0;
}


GENERATE_TEST ()
