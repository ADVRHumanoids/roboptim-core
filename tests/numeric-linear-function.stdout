Numeric linear function
  A = [5](1.2,3.4,5.6,7.8,0)
  B = 1
f(x) = 44.6
G(x) = [5](1.2,3.4,5.6,7.8,0)
H(x) = [5,5]((0,0,0,0,0),(0,0,0,0,0),(0,0,0,0,0),(0,0,0,0,0),(0,0,0,0,0))
